File: scripting/engine/Playground/Playground/Vector2.cs
namespace Playground;

public struct Vector2(float x, float y)
{
    public float X = x;
    public float Y = y;

    public static Vector2 Zero => new(0f, 0f);
    public static Vector2 One => new(1f, 1f);
    public static Vector2 Up => new(0f, 1f);
    public static Vector2 Down => new(0, -1f);
    public static Vector2 Left => new(-1f, 0);
    public static Vector2 Right => new(1f, 0);
    
    public Vector2(Vector2 vec2) : this(vec2.X, vec2.Y) {}
    public Vector2(Vector3 vec3) : this(vec3.X, vec3.Y) {}
    public Vector2(Vector4 vec4) : this(vec4.X, vec4.Y) {}
}

File: scripting/engine/Playground/math/Vector2.swift
public struct Vector2: Comparable, Codable, Equatable, Sendable {
    public let x: Float
    public let y: Float

    public var magnitude: Float {
        sqrt(x * x + y * y)
    }

    public static let zero = Vector2(x: 0, y: 0)
    public static let one = Vector2(x: 1, y: 1)
    public static let left = Vector2(x: -1, y: 0)
    public static let right = Vector2(x: 1, y: 0)
    public static let up = Vector2(x: 0, y: 1)
    public static let down = Vector2(x: 0, y: -1)

    public init(x: Float, y: Float) {
        self.x = x
        self.y = y
    }

    public static func < (lhs: Vector2, rhs: Vector2) -> Bool {
        lhs.magnitude < rhs.magnitude
    }

    public static func > (lhs: Vector2, rhs: Vector2) -> Bool {
        lhs.magnitude > rhs.magnitude
    }

    public static func ==(lhs: Vector2, rhs: Vector2) -> Bool {
        lhs.x == rhs.x && lhs.y == rhs.y
    }
}
